docs/ffi/ffi.md

```diff
@@ -0,0 +1,467 @@
+For more information about this interface, take a look at [the XLA custom call documentation](https://openxla.org/xla/custom_call).
+The full source listing can be downloaded [here](rms_norm.cc), but the key implementation details are reproduced here:
+
+```c++
+#include <functional>
+#include <numeric>
+#include <utility>
+
+#include "xla/ffi/api/c_api.h"
+#include "xla/ffi/api/ffi.h"
+
+namespace ffi = xla::ffi;
+
+std::pair<int64_t, int64_t> GetDims(ffi::Span<const int64_t> dims) {
+  if (dims.size() == 0) {
+    return std::make_pair(0, 0);
+  }
+  int64_t totalSize =
+    std::accumulate(dims.begin(), dims.end(), 1, std::multiplies<>());
+  int64_t lastDim = dims.back();
+  return std::make_pair(totalSize, lastDim);
+}
+
+ffi::Error RmsNormImpl(float eps, ffi::Buffer<ffi::DataType::F32> x,
+                       ffi::Result<ffi::Buffer<ffi::DataType::F32>> y) {
+  auto [totalSize, lastDim] = GetDims(x.dimensions);
+  if (lastDim == 0) {
+    return ffi::Error(ffi::ErrorCode::kInvalidArgument,
+                      "RmsNorm input must be an array");
+  }
+  for (int64_t n = 0; n < totalSize; n += lastDim) {
+    ComputeRmsNorm(eps, lastDim, &(x.data[n]), &(y->data[n]));
+  }
+  return ffi::Error::Success();
+}
+
+XLA_FFI_DEFINE_HANDLER(
+    RmsNorm, RmsNormImpl,
+    ffi::Ffi::Bind()
+        .Attr<float>("eps")
+        .Arg<ffi::Buffer<ffi::DataType::F32>>(/* x */)
+        .Ret<ffi::Buffer<ffi::DataType::F32>>(/* y */));
+```
+
+Starting at the bottom, we're using the XLA-provided macro `XLA_FFI_DEFINE_HANDLER` to generate some boilerplate which will expand into a function called `RmsNorm` with the appropriate signature.
+But, the important stuff here is all in the call to `ffi::Ffi::Bind()`, where we define the input and output types, and the types of any parameters.
+
+Then, in `RmsNormImpl`, we accept `ffi::Buffer` arguments which include information about the buffer shape, and pointers to the underlying data.
+In this implementation, we treat all leading dimensions of the buffer as batch dimensions, and perform RMS normalization over the last axis.
+`GetDims` is a helper function providing support for this batching behavior.
+
+### Building and registering an FFI handler
+
+Now that we have our minimal FFI wrapper implemented, we need to expose this function (`RmsNorm`) to Python.
+In this tutorial, we use [nanobind](https://nanobind.readthedocs.io) to define a tiny Python module, but it is also possible to compile `RmsNorm` into a shared library and load it using [ctypes](https://docs.python.org/3/library/ctypes.html), a pattern which we discuss below.
+
+For this example, our nanobind module is defined as follows:
+
+```c++
+#include <type_traits>
+
+#include "nanobind/nanobind.h"
+#include "xla/ffi/api/c_api.h"
+
+namespace nb = nanobind;
+
+template <typename T>
+nb::capsule EncapsulateFfiCall(T *fn) {
```

Comment:
Should we bundle this helper somewhere in jaxlib? We can implement it using Python C API without depending on nanobind.

Reply:
This is an interesting idea! My first reaction is that it doesn't really make sense to include this in `jaxlib` since it's an implementation detail and users could choose to use `ctypes` and `jex.ffi.pycapsule` if they prefer (this reminds me that I also meant to add a discussion of that to this tutorial). But, it would probably improve the developer experience if we have a clearer recommended approach. I'm inclined to leave that to future work rather than trying to overcommit with this tutorial for now.

Reply:
Drive-by comment: having one less dependency is attractive.  So out of curiosity, why not default to ctypes and leave a discussion of nanobind to future work?  What does nanobind offer here that ctypes doesn't?

Reply:
Good question! The main reason why I'm using the nanobind version here is that it lets importlib handle the logic for loading the appropriate library across platforms. I've always struggled to do this cleanly with ctypes, but it should certainly be possible!

Another minor point is that it is tricky (though not impossible) to use the ctypes approach inside of Google, and I couldn't come up with an approach that worked cleanly both internally and externally, but I'm mostly targeting external users here so it's not a huge constraint!

Reply:
Either way, I think it would be useful to present the alternatives and include a bit more discussion in this tutorial. I will do that - thanks!